Core/clim-basic/utils.lisp

```diff
@@ -743,3 +728,195 @@ a flag CLOSED is T then beginning and end of the list are consecutive too."
+                                   (breaks-rec 0 margin start end i)
+                                 (return-from breaks-rec
+                                   (values (list* start breaks)
+                                           (+ remaining-cost current-margin))))
+                               (let* ((char-breaks (%line-breaks-1 string width
+                                                                   offset margin
+                                                                   start opportunity))
+                                      (new-start (alexandria:last-elt char-breaks)))
+                                 (multiple-value-bind (breaks remaining-cost)
+                                     (breaks-rec 0 margin new-start end i)
+                                   (values (append char-breaks breaks) remaining-cost))))))
+                        (T #1=(return-from breaks-rec
+                                (values best-breaks best-cost))))
+                finally #1#)))
+    (breaks-rec initial-offset margin start end 0)))
+
+;; Implementing line breaking as defined in Unicode[1] is left as an excercise
+;; for the reader. [1] https://unicode.org/reports/tr14/ -- jd 2019-01-08
+(defun line-breaks (string width &key (break-strategy t) initial-offset margin (start 0) end)
+  "Function takes a string and returns a list of indexes where it should be split.
+
+WIDTH is a function accepting STRING, START and END arguments which should
+return string width for these boundaries. Alternatively for fixed font width it
+is a number.
+
+INITIAL-OFFSET is an initial position for the first line (may be negative). All
+remaining lines will start from the line beginning. Default is line beginning.
+
+MARGIN is a maximum width at which line should break. Defaults to 80
+characters (width of a character m is taken as a reference value).
+
+BREAK-STRATEGY may be:
+- symbol T implementing a default line breaking by word strategy,
+- symbol NIL implementing a line breaking by character strategy,
+- function accepting index which should return T for break opportunity,
+- list of characters which are break opportunities (i.e space),
+- vector of string indexes which are break opportunities.
+
+START/END denote string beginning and ending offset."
+  (unless start (setq start 0))
+  (unless end (setq end (length string)))
+  (unless margin (setq margin (* 80 (etypecase width
+                                      (function (funcall width "m" 0 1))
+                                      (number width)))))
+  (unless initial-offset (setq initial-offset 0))
+  (assert (< start end))
+  (when (null break-strategy)
+    (return-from line-breaks
+      (%line-breaks-1 string width initial-offset margin start end)))
+  (let ((opportunities (etypecase break-strategy
+                         (function
+                          (coerce (loop for i from start below end
+                                     when (funcall break-strategy i)
```

Comment:
Wouldn't it make sense to pass the string as well?

Reply:
My intention was to leave it at the caller discretion to make a closure over the string if that is necessary. From `line-breaks` perspective this interpretation makes more sense (imho).